### PR TITLE
Wiring up the UI to the events

### PR DESCRIPTION
**Pull request by chkeita**
## Summary of the Pull Request
- The UI now receives the telemetry events
- A new section for the coverage has been added 
- All synced dir are now monitored by the UI
- Gracefully exit from the UI 

depends on #663

### CODE REVIEW

**Comment by chkeita** [comment]
image

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 1]
```diff
@@ -130,40 +150,97 @@ impl TerminalUi {
...
+            match rx.recv().await {
+                Ok((event, data)) => {
+                    if let onefuzz_telemetry::Event::coverage_data = event {
+                        let (covered, features, rate) =
+                            data.iter()
+                                .cloned()
+                                .fold((None, None, None), |(c, f, r), d| match d {
+                                    EventData::Covered(value) => (Some(value), f, r),
+                                    EventData::Features(value) => (c, Some(value), r),
+                                    EventData::Rate(value) => (c, f, Some(value)),
```
Comment (chkeita) [inline src/agent/onefuzz-agent/src/local/tui.rs:219]: The rate is the value that i use to show the coverage progress bar in the UI. 
@ranweiler can you confirm that this is the intended use of that value

Comment (ranweiler) [reply src/agent/onefuzz-agent/src/local/tui.rs:219]: @chkeita, yes, that's what we want. It is the ratio of `covered / features`.

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 2]
```diff
@@ -229,6 +317,49 @@ impl TerminalUi {
...
+            let label = format!("{:.2}%", rate * 100.0);
+
+            let coverage_block = Block::default().borders(Borders::ALL).title("Coverage:");
+
+            f.render_widget(coverage_block, top_area[1]);
+
+            let gauge = Gauge::default()
+                .gauge_style(
+                    Style::default()
+                        .fg(Color::Magenta)
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/local/tui.rs:342]: Can we pick a different color scheme?  Perhaps no color scheme at all, but just the bar graph using inverted terminal defaults?  

This has accessibility issues for people who are low vision.

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 3]
```diff
@@ -221,6 +306,9 @@ impl TerminalUi {
         let file_count = ui_state.file_count;
         let mut log_event_receiver = ui_state.log_event_receiver;
         let mut terminal = ui_state.terminal;
+        let rate = ui_state.coverage.rate.unwrap_or(0.0);
+        let features = ui_state.coverage.features.unwrap_or(0);
+        let covered = ui_state.coverage.covered.unwrap_or(0);
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/local/tui.rs:311]: It appears the coverage sub-window always shows up, even if I don't have coverage enabled.

As an example, I would expect this command to not show the coverage window:

`onefuzz-agent local libfuzzer-fuzz --crashes_dir /tmp/testx/crashes/ --inputs_dir /tmp/testx/inputs/ --target_exe ~/projects/onefuzz/test-artifacts/linux-libfuzzer/fuzz.exe`

**Comment by bmc-msft** [comment]
When I run `libfuzzer-fuzz`, I don't see `inputs` as a monitored container.

**Comment by bmc-msft** [comment]
There is a ~30s delay from quitting the agent with `q` and the atexit code firing.

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 4]
```diff
@@ -130,40 +150,97 @@ impl TerminalUi {
...
+
+        while cancellation_rx.try_recv() == Err(broadcast::TryRecvError::Empty) {
+            match rx.recv().await {
+                Ok((event, data)) => {
+                    if let onefuzz_telemetry::Event::coverage_data = event {
+                        let (covered, features, rate) =
+                            data.iter()
+                                .cloned()
+                                .fold((None, None, None), |(c, f, r), d| match d {
+                                    EventData::Covered(value) => (Some(value), f, r),
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/local/tui.rs:217]: Can we include more telemetry data in what's currently listed as coverage?  Would it be difficult to make it such that it was easy to expand the available stats?

For example, when we're fuzzing, we know things like CPU Usage, Physical & Virtual Memory usage, execs/second, etc.

File: src/agent/onefuzz-agent/src/local/tui.rs [hunk 5]
```diff
@@ -229,6 +317,49 @@ impl TerminalUi {
...
                 .split(f.size());
 
+            let log_area = chunks[1];
+            let top_area = Layout::default()
+                .direction(Direction::Horizontal)
+                .constraints([Constraint::Percentage(50), Constraint::Percentage(50)].as_ref())
+                .split(chunks[0]);
+
+            let file_count_area = top_area[0];
+            let coverage_area = Layout::default()
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/local/tui.rs:327]: When I have a taller than average window (such as a 75 row terminal), the bar chart is 4 rows tall.  Is this intentional?

**Comment by bmc-msft** [comment]
If I run a libfuzzer task with files in the inputs directory, I see all of the files listed show up as a "new_coverage" event.  I would not expect these to be reported as new coverage

File: src/agent/onefuzz-agent/src/local/tui.rs (hunk 4 above)
Comment (chkeita) [reply src/agent/onefuzz-agent/src/local/tui.rs:217]: @bmc-msft  what do you thing of this 
image

File: src/agent/onefuzz-agent/src/local/generic_crash_report.rs [hunk 6]
```diff
@@ -16,11 +16,15 @@ use crate::{
 use anyhow::Result;
 use clap::{App, Arg, SubCommand};
 use storage_queue::QueueClient;
+use tokio::sync::mpsc::UnboundedSender;
+
+use super::common::UiEvent;
```
Comment (bmc-msft) [inline src/agent/onefuzz-agent/src/local/generic_crash_report.rs:21]: UiEvent lives in `crate::local::common`, which is used above.

**Review by bmc-msft — APPROVED** [review]

**Pull request merged**